qiskit_experiments/analysis/curve_fitting.py

```diff
@@ -0,0 +1,220 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Curve fitting functions for experiment analysis
+"""
+# pylint: disable = invalid-name
+
+from typing import List, Dict, Tuple, Callable, Optional
+
+import numpy as np
+import scipy.optimize as opt
+
+from qiskit.exceptions import QiskitError
+from qiskit_experiments.base_analysis import AnalysisResult
+from qiskit_experiments.analysis.data_processing import filter_data
+
+
+def curve_fit(
+    func: Callable,
+    xdata: np.ndarray,
+    ydata: np.ndarray,
+    p0: np.ndarray,
+    sigma: Optional[np.ndarray] = None,
+    **kwargs,
+) -> AnalysisResult:
+    """Use non-linear least squares to fit a function, f, to data
+
+    Wraps scipy.optimize.curve_fit
+
+    Args:
+        func: a fit function `f(x *params)`.
+        xdata: a 1D array of x-data
+        ydata: a 1D array of y-data
+        p0: initial guess for optimization parameters.
```

Comment:
While you are wrapping curve_fit, you could consider using a dict for `p0` instead of an array. I have used wrapper functions that do this for a while and not had complaints. The way what I have used works is like this:

```
def curve_fit(...

    keys = list(p0)
    def wrapper(x, p0):
        func_kwargs = {k: v for k, v in zip(keys, p0)}
        return func(x, **func_kwargs)

    popt, pcov = opt.curve_fit(wapper, xdata, ydata, sigma=sigma, p0=[p0[k] for k in keys], **kwargs)
    popt = {k: v for k, v in zip(keys, popt)}
    ...
```

The benefit is that you get to work with a dict where the parameters are labeled instead of needing to keep track of which index is for each fit parameter. A downside is that `pcov` is more awkward (a dict with tuples of keys from p0 as the keys? left as a matrix with the list giving the order of p0 keys to indices also returned?), but I have never needed to work with `pcov` personally.

Reply:
I have a few questions about this, let's discuss this more offline